### PR TITLE
formatting updates

### PR DESCRIPTION
**Pull request by dougmills-DIT**
# TP2000-761 - Crude interim fix for importer when handling missing goods nomenclature description periods on import

This fix is crude, and will be superseded by the correct separation of goods nomenclature description periods into a new table. In the interim this fix resolves a long standing issue with imports disregarding goods nomenclature description updates where the period is not also created in the TGB file. 

## Why
 * The current implementation if the importer would discard silently any record that was missing dependencies, in this case we have observed several instances where the goods nomenclature description was not applied to a workbasket from an import because the period was defined in another (previous) TGB file - that combined with the way descriptions (and periods) were implemented, it was not possible to reconcile the period and was discarded. 
 * expanding the reason the issue occured : TAP does not handle goods nomenclature description periods separately, but simply tacks on a date to the end of the description. The result is that no goods nomenclature description period SID data is recorded in TAP, and when a description update in a TGB file is encountered without an accompanying period, we cant lookup the period in the database, hence there is a missing dependency. 
 
## What
What is this PR doing, e.g. implementations, algorithms, etc.?
 *  Solution: I have added code to collect issues with imports (which will be used in the future for reporting, and possibly committed to the database) during the import process, there is a check to handle any import issues found, and a check to see if there is a method on the handler that corresponds to the issue to correct it. I have only added one such method at the moment, but the framework is there to add additional methods to perform similar operations. the method is called "create_missing_goods_nomenclature_description_period"and is found in the GoodsNomenclatureDescriptionHandler class, it's simple and just adds in a start date and removes the dependency from the description during import to allow the process to complete without dropping records. 
 * There are two edge cases that could be an issue : 
 * - When the end date of the goods nomenclature is before today, however if that's the case it does not matter since it will be historical
 * - there already exists a description with this date, which will cause a rule violation, also unlikely but possible. Importing the same TGB file the day after would resolve this issue. 
 * Unit tests cover this scenario (but we could do with a lot more around the importer)
 * Tested previously failed files, and they now work (injecting today's date as the descriptions validity_start date) 

## Checklist
- Requires migrations? No
- Requires dependency updates? No

Links to relevant material
See: [Ticket](https://uktrade.atlassian.net/browse/TP2000-761)

### CODE REVIEW

File: importer/handlers.py [hunk 1]
```diff
@@ -444,6 +495,46 @@ def build(self) -> Set[str]:
         self.dependency_keys.add(self.key)
         return self.dependency_keys
 
+    def get_import_issues(self):
+        if not self.resolve_dependencies():
+            if not self.resolve_dependencies():
```
Comment (a-gleeson) [inline importer/handlers.py:500]: double check?

File: importer/handlers.py [hunk 2]
```diff
@@ -268,13 +290,22 @@ def _generate_dependency_keys(self) -> Set[str]:
...
+                tag=dependency.tag,
+                identifying_fields=self.identifying_fields,
+                data=self.data,
             )
+
+            # using this to allow the correct mapping to be used later when reporting back to user
+            self.dependency_key_mapping[key] = {
+                "tag": dependency.tag,
+                "identifying_fields": self.identifying_fields,
+                "data": self.data,
```
Comment (paulpepper-trade) [inline importer/handlers.py:304]: Perhaps worth creating type (data class or named tuple) for this?

File: importer/handlers.py [hunk 3]
```diff
@@ -304,6 +335,26 @@ def resolve_dependencies(self) -> bool:
...
         return True
 
+    def _get_missing_dependencies(self) -> list:
+        """
+        Search the cache for all object dependencies and collect the keys and
+        return them.
+
+        Previously found objects, which are dependent on the current object,
+        should be stored in the cache. This method loops over the current
+        objects dependencies and attempts to where there is no match it will
```
Comment (paulpepper-trade) [inline importer/handlers.py:345]: "and attempts to..." missing words?

File: importer/handlers.py [hunk 4]
```diff
@@ -304,6 +335,26 @@ def resolve_dependencies(self) -> bool:
...
+        return the matching key
+        """
+        dependencies = self.dependency_keys.copy()
```
Comment (paulpepper-trade) [inline importer/handlers.py:348]: Maybe worth adding a class-level type annotation for the `dependency_keys` attribute.

File: importer/handlers.py [hunk 5]
```diff
@@ -444,6 +495,46 @@ def build(self) -> Set[str]:
...
+                            missing_dependency_data["tag"],
+                            missing_dependency_data["identifying_fields"],
+                            key,
+                            dep_missing_details,
+                        ),
+                    )
+
+                    # raise exception
+
+                    # raise Exception(f'There are missing dependencies for {self.tag} : {dep_missing_details}')
```
Comment (paulpepper-trade) [inline importer/handlers.py:531]: Redundant comments?

File: importer/taric.py [hunk 6]
```diff
@@ -221,6 +221,7 @@ class EnvelopeError(ParserError):
 class EnvelopeParser(ElementParser):
     tag = Tag("envelope", prefix=ENVELOPE)
     transaction = TransactionParser(many=True)
+    issues = list()
```
Comment (paulpepper-trade) [inline importer/taric.py:224]: It looks as though nothing in this class accesses `issues`. Is it used? If it's used from outside the class then a docstring would be useful here to aid understanding those dependencies.

File: importer/tests/taric_xml_stream/test_goods_nomenclature_description_period_scenarios.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    settings,
+):
+    settings.CACHES = {
+        "default": {
+            "BACKEND": "django.core.cache.backends.locmem.LocMemCache",
+        },
+    }
+
+    assert (
+        GoodsNomenclatureDescription.objects.all()
```
Comment (paulpepper-trade) [inline importer/tests/taric_xml_stream/test_goods_nomenclature_description_period_scenarios.py:67]: You can ditch the `.all()`.

File: importer/tests/taric_xml_stream/test_goods_nomenclature_description_period_scenarios.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    taric.process_taric_xml_stream(
+        BytesIO(goods_description_with_period_create_xml_as_text.encode()),
+        WorkBasketFactory.create().id,
+        WorkflowStatus.EDITING,
+        get_partition_scheme(settings.TRANSACTION_SCHEMA),
+        valid_user.username,
+    )
+
+    assert (
+        GoodsNomenclatureDescription.objects.all()
```
Comment (paulpepper-trade) [inline importer/tests/taric_xml_stream/test_goods_nomenclature_description_period_scenarios.py:85]: Ditch `.all()`.

File: importer/tests/taric_xml_stream/test_goods_nomenclature_description_period_scenarios.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        .filter(
+            described_goods_nomenclature__item_id="2903691100",
+            described_goods_nomenclature__suffix=80,
+        )
+        .count()
+        == 2
+    )
+
+    assert GoodsNomenclatureDescription.objects.all().filter(
```
Comment (paulpepper-trade) [inline importer/tests/taric_xml_stream/test_goods_nomenclature_description_period_scenarios.py:94]: Ditch `.all()`.

**Review by paulpepper-trade — APPROVED** [review]

File: importer/handlers.py (hunk 1 above)
Comment (dougmills-DIT) [reply importer/handlers.py:500]: ooh, didn't spot that - good catch! will update

Comment (dougmills-DIT) [reply importer/handlers.py]: updated

File: importer/handlers.py [hunk 10]
```diff
@@ -226,6 +246,8 @@ def get_link_to_model_b_link(self, model, kwargs):
     links: Iterable[LinksType] = None
     serializer_class: Type[ModelSerializer] = None
     tag: str = None
+    dependency_key_mapping = {}
```
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply importer/handlers.py]: added type (correctly called out code smell : data clump, fixed now

File: importer/handlers.py (hunk 2 above)
Comment (dougmills-DIT) [reply importer/handlers.py]: code smell data clump : updated now

File: importer/handlers.py (hunk 3 above)
Comment (dougmills-DIT) [reply importer/handlers.py:403]: updated desc, and added return value doc string.

File: importer/handlers.py (hunk 4 above)
Comment (dougmills-DIT) [reply importer/handlers.py:405]: Added done - can you check this is what you meant

File: importer/handlers.py (hunk 2 above)
Comment (dougmills-DIT) [reply importer/handlers.py]: please check this is what you were expecting

File: importer/handlers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dougmills-DIT) [reply importer/handlers.py]: please check this is as you expect

**Review by paulpepper-trade — APPROVED** [review]

**Pull request merged**